backend/orchestrators/AdvancedOrchestrator.py

```diff
@@ -0,0 +1,271 @@
+import requests
+import os
+import copy
+import json
+import logging
+import openai
+from flask import Response, request, jsonify
+from .Orchestrator import Orchestrator
+from .LoadBalancer import LoadBalancer
+
+class AdvancedOrchestrator(Orchestrator):
+    # Post chat info if data configured
+    def conversation_with_data(self, request_body, message_uuid):
+        # Create a LoadBalancer object
+        load_balancer = LoadBalancer()
+
+        # Get a weighted random OpenAIContext object
+        openai_context = load_balancer.get_openai_context()
+
+        resource = openai_context["resource"]
+        model = openai_context["model"]
+        endpoint_url = openai_context["endpoint_url"]
+        key = openai_context["key"]
+        version = openai_context["version"]
+
+        # Set up request variables
+        body, headers = self.prepare_body_headers_with_data(request, key)
+        base_url = endpoint_url if endpoint_url else f"https://{resource}.openai.azure.com/"
+        endpoint = f"{base_url}openai/deployments/{model}/extensions/chat/completions?api-version={version}"
+        history_metadata = request_body.get("history_metadata", {})
+
+        # Return response if streaming is not enabled
+        if not super().SHOULD_STREAM:
+            r = requests.post(endpoint, headers=headers, json=body)
+            status_code = r.status_code
+            r = r.json()
+
+            # Check for preview api version
+            if version == "2023-06-01-preview":
+                r['history_metadata'] = history_metadata
+                return Response(super().format_as_ndjson(r), status=status_code)
+            else:
+                result = super().formatApiResponseNoStreaming(r)
+                result['history_metadata'] = history_metadata
+                return Response(super().format_as_ndjson(result), status=status_code)
+
+        # Return response if streaming is enabled
+        else:
+            return Response(super().stream_with_data(body, headers, endpoint, message_uuid, history_metadata), mimetype='text/event-stream')
+
+    # Post chat info if data not configured
+    def conversation_without_data(self, request_body, message_uuid):
+        # Create a LoadBalancer object
+        load_balancer = LoadBalancer()
+
+        # Get a weighted random OpenAIContext object
+        openai_context = load_balancer.get_openai_context()
+
+        resource = openai_context["resource"]
+        model = openai_context["model"]
+        endpoint_url = openai_context["endpoint_url"]
+        key = openai_context["key"]
+        # version = openai_context["version"]
+
+
+        # Setup for direct query to OpenAI
+        openai.api_type = "azure"
+        openai.api_base = endpoint_url if endpoint_url else f"https://{resource}.openai.azure.com/"
+        openai.api_version = "2023-08-01-preview"
```

Comment:
Why not use what comes back from the load balancer (line 64)?

Reply:
Actually, same for the other values (except for "azure" type).

Reply:
Discussed and completed.